test/src/ui/keymap.spec.ts

```diff
@@ -0,0 +1,799 @@
+          command: 'test2'
+        });
+        let event1 = generate('keydown', { keyCode: 83, ctrlKey: true });
+        let event2 = generate('keydown', { keyCode: 68, ctrlKey: true });
+
+        expect(count1).to.be(0);
+        expect(count2).to.be(0);
+        node.dispatchEvent(event1);
+        expect(count1).to.be(0);
+        expect(count2).to.be(0);
+        node.dispatchEvent(event2);
+        expect(count1).to.be(0);
+        expect(count2).to.be(1);
+
+        binding1.dispose();
+        binding2.dispose();
+        command1.dispose();
+        command2.dispose();
+        document.body.removeChild(node);
+      });
+
+      it('should recognize permutations of modifiers', () => {
+        let node = createElement();
+        node.addEventListener('keydown', (event) => {
+          keymap.processKeydownEvent(event)
+        });
+
+        let count1 = 0;
+        let count2 = 0;
+        let command1 = commands.addCommand('test1', {
+          execute: () => { count1++; }
+        });
+        let command2 = commands.addCommand('test2', {
+          execute: () => { count2++; }
+        });
+        let binding1 = keymap.addBinding({
+          keys: ['Shift Alt Ctrl T'],
+          selector: `#${node.id}`,
+          command: 'test1'
+        });
+        let binding2 = keymap.addBinding({
+          keys: ['Alt Shift Ctrl Q'],
+          selector: `#${node.id}`,
+          command: 'test2'
+        });
+        let event1 = generate('keydown', {
+          keyCode: 84,
+          ctrlKey: true,
+          altKey: true,
+          shiftKey: true
+        });
+        let event2 = generate('keydown', {
+          keyCode: 81,
+          ctrlKey: true,
+          altKey: true,
+          shiftKey: true
+        });
+
+        expect(count1).to.be(0);
+        node.dispatchEvent(event1);
+        expect(count1).to.be(1);
+        expect(count2).to.be(0);
+        node.dispatchEvent(event2);
+        expect(count2).to.be(1);
+
+        binding1.dispose();
+        binding2.dispose();
+        command1.dispose();
+        command2.dispose();
+        document.body.removeChild(node);
+      });
+
+      it('should play back a partial match that was not completed', () => {
+        let node = createElement();
+        node.addEventListener('keydown', (event: KeyboardEvent) => {
+          keymap.processKeydownEvent(event);
+        });
+
+        let codes: number[] = [];
+        let keydown = (event: KeyboardEvent) => { codes.push(event.keyCode); };
+        document.body.addEventListener('keydown', keydown);
+
+        let called = false;
+        let command = commands.addCommand('test', {
+          execute: () => { called = true }
+        });
+        let binding = keymap.addBinding({
+          keys: ['D', 'D'],
+          selector: `#${node.id}`,
+          command: 'test'
+        });
+        // The bubbles value needs to be set explicitly, see:
+        // https://github.com/blakeembrey/simulate-event/pull/12
```

Comment:
@blink1073 FYI, I put this comment in linking to your `simulate-event` issue because it's not obvious why `bubbles: true` is included in this test but not most of the other ones.